### DISCUSSION TITLE
AttributeError: 'tuple' object has no attribute 'shape'

### DISCUSSION THREAD

**Discussion by g-romeroc** [MaartenGr/BERTopic#1463] [Q&A]
Hello, I have the following code:

```
## Import - data set
df1 = pd.read_csv('data.csv')
df1.rename(columns={'total': 'Text'}, inplace=True)

##  Train and Test Set
df1_train = df1[0:442248]
df1_test = df1[442248:542248].reset_index(drop=True, inplace=True)

train = df1_train['Text'] #Train set
test = df1_test['Text'] #Test Set


## Prepare Vocab
# Extract vocab to be used in BERTopic
vocab = collections.Counter()
tokenizer = CountVectorizer().build_tokenizer()
for doc in tqdm(tx):
    vocab.update(tokenizer(doc))
vocab = [word for word, frequency in vocab.items() if frequency >= 50]; len(vocab)

## Embedding
embedding_model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
embeddings = embedding_model.encode(train, show_progress_bar=True)





###### Model #####
%%time
#### Prepare sub-models ####
# 1. Embedding
embedding_model = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')
# 2. Dimensionality Reduction
umap_model = UMAP(n_components=5, n_neighbors=50,  random_state=42, metric="cosine", verbose=True)
# 3. Clustering
#hdbscan_model = HDBSCAN(min_samples=20, gen_min_span_tree=True, prediction_data=True, min_cluster_size=20, verbose=True, metric='euclidean', cluster_selection_method='eom')
hdbscan_model = HDBSCAN(min_samples=30, gen_min_span_tree=True, prediction_data=True, min_cluster_size=30, verbose=True)
# 4. Vectorizer
vectorizer_model = CountVectorizer(vocabulary=vocab)

# 6. Fine-tuning
representation_model = KeyBERTInspired()


# Fit BERTopic without actually performing any clustering
topic_model= BERTopic(
        embedding_model=embedding_model,
        umap_model=umap_model,
        hdbscan_model=hdbscan_model,
        vectorizer_model=vectorizer_model,
        representation_model = representation_model,
        n_gram_range=(1,3),
        calculate_probabilities = True,
        language = 'multilingual',
        nr_topics=40,
        top_n_words = 10,
        verbose=True
)



### Train ###
topics, probabilities = topic_model.fit_transform(train, embeddings=embeddings)


### Test ###
docs_test = test.to_list()
# Prever os tópicos dos novos documentos
tp, pp = topic_model.transform(docs_test)

```

Until the training, using `fit_transform()`, it worked. However, when I tried to test the model using the `transform()` function, it gave the following error:

```
2023-08-09 14:27:28,570 - BERTopic - Reduced dimensionality
[D] [14:26:31.489791] /__w/cuml/cuml/cpp/src/umap/runner.cuh:347 Running transform
[D] [14:26:31.489916] /__w/cuml/cuml/cpp/src/umap/runner.cuh:349 Building KNN Graph
[D] [14:27:28.411201] /__w/cuml/cuml/cpp/src/umap/runner.cuh:382 Smoothing KNN distances
[D] [14:27:28.425756] /__w/cuml/cuml/cpp/src/umap/runner.cuh:414 Executing fuzzy simplicial set
[D] [14:27:28.429025] /__w/cuml/cuml/cpp/src/umap/runner.cuh:443 Performing L1 normalization
[D] [14:27:28.465330] /__w/cuml/cuml/cpp/src/umap/runner.cuh:479 n_epochs=30
[D] [14:27:28.488126] /__w/cuml/cuml/cpp/src/umap/runner.cuh:502 Computing # of epochs for training each sample
[D] [14:27:28.488585] /__w/cuml/cuml/cpp/src/umap/runner.cuh:509 Performing optimization
2023-08-09 14:27:32,404 - BERTopic - Calculated probabilities with HDBSCAN
2023-08-09 14:27:32,405 - BERTopic - Predicted clusters
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[63], line 2
      1 # Prever os tópicos dos novos documentos
----> 2 tp, pp = topic_model.transform(docs_test)

File /opt/conda/lib/python3.10/site-packages/bertopic/_bertopic.py:520, in BERTopic.transform(self, documents, embeddings, images)
    517     logger.info("Predicted clusters")
    519     # Map probabilities and predictions
--> 520     probabilities = self._map_probabilities(probabilities, original_topics=True)
    521     predictions = self._map_predictions(predictions)
    522 return predictions, probabilities

File /opt/conda/lib/python3.10/site-packages/bertopic/_bertopic.py:3766, in BERTopic._map_probabilities(self, probabilities, original_topics)
   3764 # Map array of probabilities (probability for assigned topic per document)
   3765 if probabilities is not None:
-> 3766     if len(probabilities.shape) == 2:
   3767         mapped_probabilities = np.zeros((probabilities.shape[0],
   3768                                          len(set(mappings.values())) - self._outliers))
   3769         for from_topic, to_topic in mappings.items():

AttributeError: 'tuple' object has no attribute 'shape'


```

**Comment by MaartenGr** [comment]
Hmmm, I am not entirely sure what is happening here. Could you try using the latest version of cuML's UMAP? Also, which version of BERTopic and its dependencies do you have?

**Reply by wwdda** [reply]
Hi MaartenGr,

I've encountered the same issue. I've saved my model using pickle, and tried to transform new docs after loading it. But the same issue occurred.  I'm wondering if you could help me check if there's anything wrong with my implementation. Thanks a lot for your time and kindness!
The version information is:
`>>> bertopic.__version__`
`'0.15.0'`
`>>> import cuml`
`>>> cuml.__version__`
`'23.08.00'`

The code and error are as below:
`from bertopic import BERTopic`
`import json`
`import pandas as pd`
`loaded_model = BERTopic.load('models/self_disclosure_models/self_disclosure_n10_num25_pickle_model')`
`corpus = pd.read_excel('predict_corpus.xlsx')`
`docs = corpus['text'].astype('string').tolist()`
`loaded_model.transform(docs)`

/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
Batches: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1786/1786 [00:33<00:00, 53.27it/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/bertopic/_bertopic.py", line 520, in transform
    probabilities = self._map_probabilities(probabilities, original_topics=True)
  File "/home/dvrkdemo/miniconda3/envs/bertopic/lib/python3.10/site-packages/bertopic/_bertopic.py", line 3766, in _map_probabilities
    if len(probabilities.shape) == 2:
AttributeError: 'tuple' object has no attribute 'shape'

**Reply by wwdda** [reply]
Besides, I also tried loaded_model.fit_transform(docs), this will cause segmentation fault(core dumped)...

and I also tried on the trained dataset, but the same error still occurs.

**Reply by MaartenGr** [reply]
Strange, I am not quite sure what is happening here. Instead, I would advise saving with `safetensors` instead. That will skip the usage of the clustering algorithm which should prevent the issue from occurring.
